kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,352 @@
+ *  This class also provides interface to access the data of the constraint equations.
+ *  
+ */
+class ConstraintEquationContainer
+{
+
+  public:
+    /// Pointer definition of ConstraintEquationContainer
+    KRATOS_CLASS_POINTER_DEFINITION(ConstraintEquationContainer);
+
+    typedef Dof<double> DofType;
+    typedef size_t IndexType;
+    typedef Node<3> NodeType;
+    typedef ConstraintEquation::Pointer ConstraintEquationPointerType;
+
+    struct EquationHasher
+    {
+        size_t
+        operator()(const ConstraintEquationPointerType &rObj) const
+        {
+            std::size_t seed = 0;
+            boost::hash_combine(seed, rObj->SlaveDofId());
+            boost::hash_combine(seed, rObj->SlaveDofKey());
+            return seed;
+        }
+    };
+
+    // Custom comparator that compares the string objects by length
+    struct EquationComparator
+    {
+        bool
+        operator()(const ConstraintEquationPointerType &rObj1, const ConstraintEquationPointerType &rObj2) const
+        {
+            if ((rObj1->SlaveDofId() == rObj2->SlaveDofId()) && (rObj1->SlaveDofKey() == rObj2->SlaveDofKey()))
+                return true;
+            return false;
+        }
+    };
+
+    typedef std::unordered_set<ConstraintEquationPointerType, EquationHasher, EquationComparator> ConstraintEquationVectorType;
+
+    typedef ConstraintEquationVectorType::iterator iterator;
+
+    ///@name Life Cycle
+    ///@{
+
+    /**
+		Creates a MPC data object
+		*/
+    ConstraintEquationContainer() : mConstraintEquationsVector()
+    {
+    }
+    /// Destructor.
+    virtual ~ConstraintEquationContainer(){};
+
+    ///@}
+
+    ///@name Access
+    ///@{
+
+    /**
+		Clears the maps contents
+		*/
+    void Clear()
+    {
+        mConstraintEquationsVector.clear();
+    }
+    iterator begin() { return mConstraintEquationsVector.begin(); }
+    iterator end() { return mConstraintEquationsVector.end(); }
+
+    /**
+		Get the Constraint equation for this slave
+		@return Data vector for this slave
+		*/
+    ConstraintEquation &GetConstraintEquation(DofType &rSlaveDof)
+    {
+        ConstraintEquationPointerType equation = Kratos::make_shared<ConstraintEquation>(rSlaveDof);
+        auto res = mConstraintEquationsVector.find(equation);
+        if (res != mConstraintEquationsVector.end())
+        {
+            return *(*res);
+        }
+        else
+        {
+            KRATOS_THROW_ERROR("Runtime error :: ","No constraint equation found for given slave dof.\nThere is something worng","");
+        }
+    }
+
+    /**
+		Get the Total number of MasterDOFs for a given slave dof
+		@return Total number of MasterDOFs for a given slave dof
+		 */
+    int GetNumbeOfMasterDofsForSlave(DofType const &rSlaveDof)
+    {
+        ConstraintEquationPointerType equation = Kratos::make_shared<ConstraintEquation>(rSlaveDof);

```

Comment:
It is needed for ".find" operation on the set. as it is a set of  ConstraintEquationPointerType and so we need a ConstraintEquationPointerType to find in the set.
What I can do is combine the lines. But a shared pointer is necessary because of the set type.